Task 4

IvanPavlov90 opened the pull request.

IvanPavlov90 committed:
starting task 4

- initialize objects for menu, recording data, following changes
- add serializtor for json
- add log files

IvanPavlov90 committed:
add classes for reading json log and data validation for user input.

- add special class for reading json log
- add class for build files in folder based on information from log
- add method for validation user data input
- remade method that is responsible for writing information in log

IvanPavlov90 committed:
create builder class

- this class is responsible for creating folder's state on certain date
- make some corrections in my classes

IvanPavlov90 committed:
final corrections before PR

- fix deserialization of private fields
- add relative ways to folder and log file
- add rewriting of log file when you start another session of program

IvanPavlov90 committed:
add create directory method for case if observable directory doesn't exist

IvanPavlov90 committed:
23.04.2021 fix

- create method for correct clearing of subfolders, now all structure that was created by user saves as empty folders.

STeeL835 commented on Task 4/Log.json:
```diff
@@ -0,0 +1 @@
+﻿
```
Empty file in the repository?

STeeL835 commented on Task 4/Task 4/Task 4/AppMenu.cs:
```diff
@@ -0,0 +1,72 @@
+            RollingChanges = 2
+        }
+
+        public static void ChosingMode ()
```
I think you meant `ChooseMode` or `Start` (menu start)

STeeL835 commented on Task 4/Task 4/Task 4/AppMenu.cs:
```diff
@@ -0,0 +1,72 @@
+            {
+                string directoryPath = GetDirectory() + @"\Files";
+                CreateDirectory(directoryPath);
+                string logPath = GetDirectory() + @"\Log.json";
```
What if file doesn't exists?

STeeL835 commented on Task 4/Task 4/Task 4/AppMenu.cs:
```diff
@@ -0,0 +1,72 @@
+        {
+            string path = Directory.GetParent(Directory.GetCurrentDirectory()).Parent.Parent.Parent.Parent.FullName;
+            return path;
+        }
```
Which directory is this? (Method name)
Which directory is this? (`GetParent.Parent.Parent.Parent.Parent`)
I'm going to launch this in my root folder for example. What will happen?

You could simply ask user which folder they want to monitor

STeeL835 commented on Task 4/Task 4/Task 4/Print.cs:
```diff
@@ -0,0 +1,16 @@
+        {
+            Console.WriteLine(message);
+        }
+    }
```
This wrapper is useless - it's also static, and doesn't add any logic.

STeeL835 commented on Task 4/Task 4/Task 4/InputHelper.cs:
```diff
@@ -0,0 +1,35 @@
+                    string format = "dd.MM.yyyy HH:mm:ss";
+                    var dateTime = DateTime.ParseExact(value, format, provider);
+                    return dateTime;
+                } while (value.Trim().Length <= 0);
```
* `!string.IsNullOrWhitespace()`
* practically useless - you check this after trying to parse a date, which will fail anyway if value is whitespace

STeeL835 commented on Task 4/Task 4/Task 4/AppMenu.cs:
```diff
@@ -0,0 +1,72 @@
+                        do
+                        {
+                            date = InputHelper.InputDate();
+                        } while (date == defaultDate);
```
InputDate should take care of providing a correct value. Now the date picking logic exists in several classes, when it really shouldn't.
Also, for that matter, you _will_ forget what default date is. Next time you need a data input, you will remember the wrong one and will be surprised, when this piece of code won't work as expected. This is a code smell that leads to bugs.

1. Such values should be made a constant or a static readonly value that can be accessible if needed
2. InputDate tells user to try again, instead of retry - it returns some magic value. Make it handle this and remove while from here.

STeeL835 replied:
Move cases into methods here

STeeL835 commented on Task 4/Task 4/Task 4/AppMenu.cs:
```diff
@@ -0,0 +1,72 @@
+            DirectoryInfo dir = new DirectoryInfo(path);
+            if (!dir.Exists)
+                dir.Create();
+        }
```
[`Directory.CreateDirectory`](https://docs.microsoft.com/en-us/dotnet/api/system.io.directory.createdirectory) returns a directory if it already exists

STeeL835 commented on Task 4/Task 4/Task 4/Builder.cs:
```diff
@@ -0,0 +1,75 @@
+
+namespace Task_4
+{
+    public static class Builder
```
Name's too generic. Try to reflect its purpose in the class name, like `FolderStateBuilder` (or -manager)

STeeL835 commented on Task 4/Task 4/Task 4/FileEventsInfo.cs:
```diff
@@ -0,0 +1,64 @@
+                throw new ArgumentException("You can't put empty null here.");
+            }
+            return str;
+        }
```
If you catch such exception you won't be able to figure out what 'here' means, and which parameter isn't supposed to be empty.

well, you will, if you check stack trace (if you have it) and look at the code to find a problematic line, but if you don't have access to a source code - your only chance is to bruteforce all parameter combinations to learn what's going on

STeeL835 commented on Task 4/Task 4/Task 4/FileEventsInfo.cs:
```diff
@@ -0,0 +1,64 @@
+
+        private string CheckEmptyOrNull (string str)
+        {
+            if (str == String.Empty || str == null)
```
`string.NullOrEmpty`/`string.NullOrWhitespace`

STeeL835 commented on Task 4/Task 4/Task 4/FileLog.cs:
```diff
@@ -0,0 +1,30 @@
+            using (StreamWriter writer = new StreamWriter(logPath, true, System.Text.Encoding.UTF8))
+            {
+                  writer.WriteLine(Serialize(log));
+            }
```
You could just put a serialized string into WriteAllText. Why double work?

STeeL835 commented on Task 4/Task 4/Task 4/FileWatcher.cs:
```diff
@@ -0,0 +1,73 @@
+                                 | NotifyFilters.FileName
+                                 | NotifyFilters.CreationTime
+                                 | NotifyFilters.LastWrite
+                                 | NotifyFilters.LastAccess;
```
Are you able to restore file attributes? If no, then we probably shouldn't track attribute changes.
LastAccess attribute isn't related to changes

STeeL835 commented on Task 4/Task 4/Task 4/FileWatcher.cs:
```diff
@@ -0,0 +1,73 @@
+            DateTime date = DateTime.Now;
+            FileEventsInfo fileEventInfo = new FileEventsInfo(fileEvent.FullPath, "", date, "Delete", "");
+            AddFileEventInfoToLog(fileEventInfo);
+        }
```
File moving triggers Delete and Create events. If I move files, restore doesn't restore their content

STeeL835 commented on Task 4/Task 4/Task 4/FileEventsInfo.cs:
```diff
@@ -0,0 +1,64 @@
+        public FileEventsInfo(string path, string oldPath, DateTime changes, string typeOfEvent, string content)
+        {
+            FullPath = CheckEmptyOrNull(path);
+            OldFullPath = CheckStrNull(oldPath);
```
if file was just created it's reasonable that it doesn't have oldPath. Forcing yourself to use everything but null to say that there's actually no value - is forcing yourself to use crutches

STeeL835 commented on Task 4/Task 4/Task 4/FileEventsInfo.cs:
```diff
@@ -0,0 +1,64 @@
+        public DateTime LastChangesTime { get; private set; }
+
+        [JsonInclude]
+        public string EventType { get; private set; }
```
I know that c# has a specific Type structure that is perfect for such thing as enumerations like types of something... And serializer should be able to serialize it

STeeL835 commented on Task 4/Task 4/Task 4/Builder.cs:
```diff
@@ -0,0 +1,75 @@
+    {
+        public static void BuildState (List<FileEventsInfo> fileEvent, DateTime date, string directoryPath)
+        {
+            ClearDirectory(directoryPath);
```
Noo, my precious files that were created before I started observation mode!

You basically need to save an initial state on start of monitoring and rollback to it if you want to apply changes from the start.

STeeL835 commented on Task 4/Task 4/Task 4/Builder.cs:
```diff
@@ -0,0 +1,75 @@
+            string path = file.FullPath;
+            switch (file.EventType)
+            {
+                case "Create":
```
One mistake and you've mistaken. If only there was a convenient type structure for options that could use less memory and compiler could notice that the value is wrong..

STeeL835 commented on Task 4/Task 4/Task 4/Builder.cs:
```diff
@@ -0,0 +1,75 @@
+                        writer.WriteLine(file.Content);
+                    }
+                    File.SetLastWriteTime(path, file.LastChangesTime);
+                    break;
```
I deleted "dad" and wrote "mom" in the file. After restore the file contained both lines, which is not the state that was at the time

STeeL835 commented on Task 4/Task 4/Task 4/Builder.cs:
```diff
@@ -0,0 +1,75 @@
+                        writer.WriteLine(file.Content);
+                    }
+                    File.SetLastWriteTime(path, file.LastChangesTime);
+                    break;
```
In system terms, **renaming - is moving**. No matter if you are renaming the file, the folder it is in or moving it to another folder - you change file's path. 
I don't know why FileSystemWatcher has a different opinion on that, but to rename a file you "move" it from "a.txt" to "b.txt"

STeeL835 commented on Task 4/Task 4/Task 4/Reader.cs:
```diff
@@ -0,0 +1,32 @@
+
+namespace Task_4
+{
+    public static class Reader
```
A bookworm, I see.

I think you should move methods from Reader and Writer to some `LogsSerializer` that would read and write this log file. Better if it is not static (to set log file name in ctor and just call methods)

ReadContent in this class is the same as File.ReadAllText()

STeeL835 commented on Task 4/Task 4/Task 4/InputHelper.cs:
```diff
@@ -0,0 +1,35 @@
+                    value = Console.ReadLine();
+                    CultureInfo provider = CultureInfo.InvariantCulture;
+                    string format = "dd.MM.yyyy HH:mm:ss";
+                    var dateTime = DateTime.ParseExact(value, format, provider);
```
You can use TryParseExact instead of catching exceptions

IvanPavlov90 committed:
fix 24.04.2021

- delete empty log file and rename start method;
- add new method that asks user to specify folder he wants to observe;
- add method that creates backup of observind folder;

IvanPavlov90 committed:
fix 25.04.2021

- replace case logic to methods in builder calss
- fix change method, now when you delete text from file and write another
both texts (previous and recent) will not show
- build group of methods that create backup of observing folder
- build method that saves content of removing files

IvanPavlov90 committed:
fix 26.04.2021

- add new validation for date;
- change name of Builder class;
- remove redundant Notify.Filters;
- added enum to FileEventsLog class just not to put strings into switch;
- added new non-static class for recording into log and reading from it;

IvanPavlov90 committed:
changes 27.04.2021

- create method for creating backup folder and log file;
- delete print class;
- make some changes with exceptions in fileeventsinfo class;

IvanPavlov90 committed:
delete reader class;

IvanPavlov90 committed:
bugfix with permanent fail

STeeL835 commented:
don't title commits `fix 24.04.2021`. Either summarize the changes or make smaller commits

STeeL835 commented on Task 4/Task 4/Task 4/FileEventsInfo.cs:
```diff
@@ -1,6 +1,7 @@
 ﻿using System;
 using System.Text.Json;
 using System.Text.Json.Serialization;
+using static Task_4.FileWatcher;
```
Better move that enum outside FileWatcher class

STeeL835 commented on Task 4/Task 4/Task 4/FileEventsInfo.cs:
```diff
@@ -30,34 +31,29 @@ public FileEventsInfo(string path, string oldPath, DateTime changes, string type
-            {
-                throw new ArgumentException("You can't put empty string or null here.");
-            }
+            if (String.IsNullOrWhiteSpace(str))
+                throw new ArgumentException(String.Format("You can't put null or white spaces into {0}", FullPath));
```
```cs
throw new ArgumentException($"You can't put null or white spaces into {FullPath}");
```
This is more readable and also in this case converts to concatenation on compile (faster than Format).

STeeL835 commented on Task 4/Task 4/Task 4/FileWatcher.cs:
```diff
@@ -6,20 +6,27 @@ namespace Task_4
         {
-            if (folderPath == null || logPath == null || folderPath == String.Empty || logPath == String.Empty)
+            if (folderPath == null ||folderPath == String.Empty)
                 throw new IOException("Path to file or folder can't be null or empty");
```
That's an argument exception, not IOException - a validation error, not an error related to input-output failures

STeeL835 commented on Task 4/Task 4/Task 4/FolderStateBuilder.cs:
```diff
@@ -0,0 +1,170 @@
+                    {
+                        FileCreation(file);
+                    }
+                    catch
+                    {
+                        CreateSubFolders(file.FullPath);
+                        FileCreation(file);
+                    }
```
Specify exception type that you expect to catch

STeeL835 commented on Task 4/Task 4/Task 4/FolderStateBuilder.cs:
```diff
@@ -0,0 +1,170 @@
+        /// It is necessary, because creating empty subfolder is not tracked, and doesn't 
+        /// record in log.
+        /// </summary>
+        private static void CreateSubFolders(string path)
+        {
+            string[] separateString = path.Split('\\');
+            string subFolderPath = String.Empty;
+            for (int i = 0; i < separateString.Length - 1; i++)
+            {
+                subFolderPath += separateString[i];
+                Directory.CreateDirectory(subFolderPath);
+                subFolderPath += @"\";
+            }
+        }
```
Docs say that Directory.CreateDirectory creates all subfolders in the path if they weren't created. So you can just pass a full path without filename there (look for a method that removes it in Path class)

STeeL835 commented on Task 4/Task 4/Task 4/FolderStateBuilder.cs:
```diff
@@ -0,0 +1,170 @@
+        /// This method creates files with the help of information from example of
+        /// FileEventsInfo class.
+        /// </summary>
+        private static void FileCreation (FileEventsInfo file)
```
Use 1st form of verbs, like `CreateFile`

STeeL835 commented on Task 4/Task 4/Task 4/FolderStateBuilder.cs:
```diff
@@ -0,0 +1,170 @@
+        /// This method contains logic of renaiming file.
+        /// </summary>
+        /// <param name="file"></param>
+        private static void FileRenaiming(FileEventsInfo file)
+        {
+            File.Delete(file.OldFullPath);
+            using (var stream = File.Create(file.FullPath)) { };
+            using (StreamWriter writer = new StreamWriter(file.FullPath, true, System.Text.Encoding.UTF8))
+            {
+                writer.WriteLine(file.Content);
+            }
+            File.SetLastWriteTime(file.FullPath, file.LastChangesTime);
+        }
```
If it's only rename, you can use File.Move to "move" file to a new name. Then set last write time

IvanPavlov90 committed:
replace enum from fileWatcher class to namespace and fix message in argument exception in fileEventsInfo class.

IvanPavlov90 committed:
fix exception type in FileWatcher class and specify exception type in FolderStateBuilder class logic

IvanPavlov90 committed:
update logic in createSubfolders method

IvanPavlov90 committed:
rename methods in FolderStateBuilder and FileWatcher class also change FileRename method in FolderStateBuilder class
